lib/source-map.js

```diff
@@ -362,22 +361,54 @@ SourceMap.prototype._scanMappings = function(srcMap, sourcesOffset, namesOffset,
   this.content.mappings = mappings;
 };
 
-SourceMap.prototype.end = function() {
-  if (this.mapCommentType === 'line') {
-    this.stream.write('//# sourceMappingURL=' + this.mapURL);
-  } else {
-    this.stream.write('/*# sourceMappingURL=' + this.mapURL + ' */');
-  }
-  mkdirp.sync(path.dirname(this.mapFile));
-  fs.writeFileSync(this.mapFile, JSON.stringify(this.content));
+SourceMap.prototype.end = function(cb, thisArg) {
+  var stream = this.stream;
+  var sourceMap = this;
+
+
   return new RSVP.Promise(function(resolve, reject) {
-    this.stream.on('finish', resolve);
-    this.stream.on('error', reject);
-    this.stream.end();
-  }.bind(this));
+    stream.on('error', function(error) {
+      stream.on('close', function() {
+        reject(error);
+      });
+
+      stream.end();
+    });
+
+    var error, success;
+
+    try {
+      if (cb) {
+        cb.call(thisArg, sourceMap);
+      }
+
+      if (sourceMap.mapCommentType === 'line') {
+        stream.write('//# sourceMappingURL=' + sourceMap.mapURL);
+      } else {
+        stream.write('/*# sourceMappingURL=' + sourceMap.mapURL + ' */');
+      }
+
+      mkdirp.sync(path.dirname(sourceMap.mapFile));
+      fs.writeFileSync(sourceMap.mapFile, JSON.stringify(sourceMap.content));
+      success = true;
+    } catch(e) {
+      success = false;
+      error = e;
+    } finally {
+      stream.on('close', function() {
+        if (success) {
+          resolve();
```

Comment:
Just a note, but if [`stream.on('finish'`](https://github.com/ef4/fast-sourcemap-concat/pull/19/files#diff-cbf7cad5c2d14ff5f8c1850747b7855bL374) above had any arguments, they would be lost here.